File: httpcsharp/httpcsharp/routes/BookRoute.cs
using System;
using System.Text;
using Grapevine.Interfaces.Server;
using Grapevine.Shared;
using httpcsharp.models;
using httpcsharp.services.data;
using httpcsharp.services.json;

namespace httpcsharp.routes
{ 
    public class BookRoute
    {
        private readonly IDataService _dataService;
        private readonly IJsonService _jsonService;

        public BookRoute(IDataService dataService, IJsonService jsonService)
        {
            _dataService = dataService;
            _jsonService = jsonService;
        }
               
        public IHttpContext Handler(IHttpContext context)
        {
            return GetHandler(context);
        }

        private IHttpContext GetHandler(IHttpContext context)
        {
            context.Response.ContentType = ContentType.JSON;

            String json;

            try
            {
                var idValues = context.Request.QueryString.GetValues("id");
                var authorIdValues = context.Request.QueryString.GetValues("author-id");
                var genreIdValues = context.Request.QueryString.GetValues("genre-id");
                
                if (idValues != null && idValues.Length > 0)
                {
                    var id = idValues[0];
                    var book = _dataService.GetBookByBookId(id);
                    json = _jsonService.Write(book);
                }
                else if (authorIdValues != null && authorIdValues.Length > 0)
                {
                    var authorId = authorIdValues[0];
                    var books = _dataService.GetBooksByAuthorId(authorId);
                    json = _jsonService.Write(books);
                }
                else if (genreIdValues != null && genreIdValues.Length > 0)
                {
                    var genreId = genreIdValues[0];
                    var books = _dataService.GetBooksByGenreId(genreId);
                    json = _jsonService.Write(books);
                }
                else
                {
                    var books = _dataService.GetBooks();
                    json = _jsonService.Write(books);
                }
            }
            catch (Exception e)
            {
                context.Response.StatusCode = HttpStatusCode.InternalServerError;
                json = _jsonService.Write(new ErrorMessage(e.Message));
            }
            
            context.Response.SendResponse(Encoding.ASCII.GetBytes(json));
            return context;
        }
    }
}

File: httpjava/src/main/java/org/stuartaroth/httpjava/routes/BookRoute.java
package org.stuartaroth.httpjava.routes;

import org.eclipse.jetty.http.HttpStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.stuartaroth.httpjava.models.Book;
import org.stuartaroth.httpjava.models.ErrorMessage;
import org.stuartaroth.httpjava.models.SuccessMessage;
import org.stuartaroth.httpjava.services.data.DataService;
import org.stuartaroth.httpjava.services.json.JsonService;
import spark.Request;
import spark.Response;

import java.util.Set;

public class BookRoute {
    private static Logger logger = LoggerFactory.getLogger(BookRoute.class);

    private JsonService jsonService;
    private DataService dataService;

    public BookRoute(JsonService jsonService, DataService dataService) {
        this.jsonService = jsonService;
        this.dataService = dataService;
    }

    public Object get(Request request, Response response) throws Exception {
        response.type("application/json");

        try {
            Set<String> queryParams = request.queryParams();
            if (queryParams.contains("id")) {
                String id = request.queryMap("id").value();
                return jsonService.write(dataService.getBookByBookId(id));
            } else if (queryParams.contains("author-id")) {
                String authorId = request.queryMap("author-id").value();
                return jsonService.write(dataService.getBooksByAuthorId(authorId));
            } else if (queryParams.contains("genre-id")) {
                String genreId = request.queryMap("genre-id").value();
                return jsonService.write(dataService.getBooksByGenreId(genreId));
            } else {
                return jsonService.write(dataService.getBooks());
            }

        } catch (Exception e) {
            logger.error("Exception:", e);
            response.status(HttpStatus.INTERNAL_SERVER_ERROR_500);
            return jsonService.write(new ErrorMessage(e.getMessage()));
        }
    }

    public Object post(Request request, Response response) throws Exception {
        response.type("application/json");
        try {
            Book book = jsonService.read(request.body(), Book.class);
            Book createdBook = dataService.createBook(book);
            return jsonService.write(createdBook);
        } catch (Exception e) {
            logger.error("Exception:", e);
            response.status(HttpStatus.INTERNAL_SERVER_ERROR_500);
            return jsonService.write(new ErrorMessage(e.getMessage()));
        }
    }

    public Object put(Request request, Response response) throws Exception {
        response.type("application/json");
        try {
            Book book = jsonService.read(request.body(), Book.class);
            Book createdBook = dataService.updateBook(book);
            return jsonService.write(createdBook);
        } catch (Exception e) {
            logger.error("Exception:", e);
            response.status(HttpStatus.INTERNAL_SERVER_ERROR_500);
            return jsonService.write(new ErrorMessage(e.getMessage()));
        }
    }

    public Object delete(Request request, Response response) throws Exception {
        response.type("application/json");
        try {
            Set<String> queryParams = request.queryParams();

            if (queryParams.contains("id")) {
                String id = request.queryMap("id").value();
                dataService.deleteBook(id);
                return jsonService.write(new SuccessMessage(String.format("Deleted `book` with id: %s", id)));
            } else {
                response.status(HttpStatus.UNPROCESSABLE_ENTITY_422);
                return jsonService.write(new ErrorMessage("You must provide an `id`"));
            }
        } catch (Exception e) {
            logger.error("Exception:", e);
            response.status(HttpStatus.INTERNAL_SERVER_ERROR_500);
            return jsonService.write(new ErrorMessage(e.getMessage()));
        }
    }
}
